waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -0,0 +1,168 @@
+    state: ChaChaPolyCipherState,
+    plaintext: openArray[byte]): ChaChaPolyCiphertext
+    {.noinit, raises: [Defect].} =
+  #TODO: add padding
+  result.data.add plaintext
+  ChaChaPoly.encrypt(state.k, state.nonce, result.tag, result.data, state.ad)
+
+proc decrypt*(
+    state: ChaChaPolyCipherState, 
+    ciphertext: ChaChaPolyCiphertext): seq[byte]
+    {.raises: [Defect, NoiseDecryptTagError].} =
+  var
+    tagIn = ciphertext.tag
+    tagOut: ChaChaPolyTag
+  result = ciphertext.data
+  ChaChaPoly.decrypt(state.k, state.nonce, tagOut, result, state.ad)
+  #TODO: add unpadding
+  trace "decrypt", tagIn = tagIn.shortLog, tagOut = tagOut.shortLog, nonce = state.nonce
+  if tagIn != tagOut:
+    debug "decrypt failed", result = shortLog(result)
+    raise newException(NoiseDecryptTagError, "decrypt tag authentication failed.")
+
+
+proc randomChaChaPolyCipherState*(rng: var BrHmacDrbgContext): ChaChaPolyCipherState =
+  brHmacDrbgGenerate(rng, result.k)
+  brHmacDrbgGenerate(rng, result.nonce)
+  result.ad = newSeq[byte](32)
+  brHmacDrbgGenerate(rng, result.ad)
+
+
+#################################################################
+
+
+# Utility
+
+proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
+  result.privateKey = Curve25519Key.random(rng)
+  result.publicKey = result.privateKey.public()
+
+
+# Public keys serializations/encryption
+
+proc `==`(k1, k2: NoisePublicKey): bool =
+  result = (k1.flag == k2.flag) and (k1.pk == k2.pk)
+  
+
+proc keyPairToNoisePublicKey*(keyPair: KeyPair): NoisePublicKey =
+  result.flag = 0
+  result.pk = getBytes(keyPair.publicKey)
+
+
+proc genNoisePublicKey*(rng: var BrHmacDrbgContext): NoisePublicKey =
+  let keyPair: KeyPair = genKeyPair(rng)
+  result.flag = 0
+  result.pk = getBytes(keyPair.publicKey)
+
+proc serializeNoisePublicKey*(noisePublicKey: NoisePublicKey): seq[byte] =
+  result.add noisePublicKey.flag
+  result.add noisePublicKey.pk
+
+#TODO: strip pk_auth if pk not encrypted
+proc intoNoisePublicKey*(serializedNoisePublicKey: seq[byte]): NoisePublicKey =
+  result.flag = serializedNoisePublicKey[0]
+  assert result.flag == 0 or result.flag == 1
+  result.pk = serializedNoisePublicKey[1..<serializedNoisePublicKey.len]
+
+# Public keys encryption/decryption
+
+proc encryptNoisePublicKey*(cs: ChaChaPolyCipherState, noisePublicKey: NoisePublicKey): NoisePublicKey
+  {.raises: [Defect, NoiseNonceMaxError].} =
+  if cs.k != EmptyKey and noisePublicKey.flag == 0:
+    let enc_pk = encrypt(cs, noisePublicKey.pk)
+    result.flag = 1
+    result.pk = enc_pk.data
+    result.pk.add enc_pk.tag
+  else:
+    result = noisePublicKey
+
+
+proc decryptNoisePublicKey*(cs: ChaChaPolyCipherState, noisePublicKey: NoisePublicKey): NoisePublicKey
+  {.raises: [Defect, NoiseDecryptTagError].} =
+  if cs.k != EmptyKey and noisePublicKey.flag == 1:
+    #let ciphertext = ChaChaPolyCiphertext(data: noisePublicKey.pk, tag: noisePublicKey.pk_auth)
+    let pk_len = noisePublicKey.pk.len - ChaChaPolyTag.len
```

Comment:
Would be easier to understand what is going on with the help of some comments!

Reply:
Addressed in https://github.com/status-im/nim-waku/commit/704f3fe40fb4ad63212ec5f6face687b4971c909